crates/common/consensus/lean/src/state.rs

```diff
@@ -48,126 +59,362 @@ impl LeanState {
+
+        // Loop each root and reconstruct the justifications from the flattened BitList
+        for (i, root) in self.justifications_roots.iter().enumerate() {
+            let mut justifications = BitList::with_capacity(VALIDATOR_REGISTRY_LIMIT as usize)
+                .map_err(|err| anyhow!("Failed to create BitList for justifications: {err:?}"))?;
+
+            // Loop each validator and set their justification
+            self.justifications_validators
+                .iter()
+                .skip(i * VALIDATOR_REGISTRY_LIMIT as usize)
+                .take(VALIDATOR_REGISTRY_LIMIT as usize)
+                .enumerate()
+                .try_for_each(|(validator_index, justification)| -> anyhow::Result<()> {
+                    justifications
+                        .set(validator_index, justification)
+                        .map_err(|err| anyhow!("Failed to set justification: {err:?}"))?;
+                    Ok(())
+                })?;
 
-    pub fn initialize_justifications_for_root(&mut self, root: &B256) -> anyhow::Result<()> {
-        if self.justifications_roots.contains(root) {
-            return Ok(());
+            // Insert the root and its justifications into the map
+            justifications_map.insert(*root, justifications);
         }
 
-        self.justifications_roots
-            .push(*root)
-            .map_err(|err| anyhow!("Failed to insert root into justifications_roots: {err:?}"))?;
-
-        let old_length = self.justifications_roots_validators.len();
-        let new_length = old_length + VALIDATOR_REGISTRY_LIMIT as usize;
+        Ok(justifications_map)
+    }
 
-        let mut new_justifications_roots_validators = BitList::with_capacity(new_length)
-            .map_err(|err| anyhow!("Failed to initialize new justification bits: {err:?}"))?;
+    /// Saves a map of `root -> justifications` back into the state's flattened
+    /// data structure.
+    pub fn set_justifications(
+        &mut self,
+        justifications_map: HashMap<B256, BitList<U4096>>,
+    ) -> anyhow::Result<()> {
+        let mut justifications_roots = VariableList::<B256, U262144>::empty();
+        let mut justifications_validators = BitList::with_capacity(
+            (MAX_HISTORICAL_BLOCK_HASHES * VALIDATOR_REGISTRY_LIMIT) as usize,
+        )
+        .map_err(|err| {
+            anyhow!("Failed to create BitList for justifications_validators: {err:?}")
+        })?;
 
-        for (i, bit) in self.justifications_roots_validators.iter().enumerate() {
-            new_justifications_roots_validators
-                .set(i, bit)
-                .map_err(|err| {
-                    anyhow!("Failed to initialize justification bits to existing values: {err:?}")
+        let mut sorted_roots = justifications_map.keys().collect::<Vec<_>>();
+        sorted_roots.sort();
+
+        for (root_index, root) in sorted_roots.iter().enumerate() {
```

Comment:
```suggestion
        for (root_index, root) in justifications_map.keys().sorted().enumerate() {
```

We can use `itertools` here for cleaner and readable code. You might have to import like:

```rust
use itertools::Itertools;
```